tarantool/connection.py

```diff
@@ -767,12 +770,30 @@ def _send_request_wo_reconnect(self, request):
 
         assert isinstance(request, Request)
 
+        # Flag for detecting the last message as out-of-band.
+        iproto_chunk_detected = False
+        
         response = None
         while True:
             try:
-                self._socket.sendall(bytes(request))
+                if not iproto_chunk_detected:
+                    # If the last received message is out-of-band, 
+                    # the request will not be sent.
+                    self._socket.sendall(bytes(request))
```

Comment:
It will be more clear:
```suggestion
                if not request_sended:
                    # If the last received message is out-of-band, 
                    # the request will not be sent.
                    self._socket.sendall(bytes(request))
                    request_sended = True
```
or even extract the send from the loop.

Reply:
AFAIU, calling the `self._socket.sendall`  in this loop may be required several times, therefore extracting `self._socket.sendall` from loop is a problem

Reply:
+1, to be honest, `iproto_chunk_detected` is rather hard to catch on from the first glance.

Reply:
To be honest, I don't see such opportunity

Reply:
Tarantool checks schema version on commit:

https://github.com/tarantool/tarantool/blob/master/src/box/iproto.cc#L2000

So it may cause the error after IPROTO_CHUNK. You could try to reproduce it with:

1. 1 terminal:

```Lua
box.schema.create_space(
 'tester', {
  format = {
    {name = 'id', type = 'unsigned'},
    {name = 'name', type = 'string'},
  }
})
box.space.tester:create_index(
 'primary_index', {
  parts = {
    {field = 1, type = 'unsigned'},
  }
})
function on_replace_callback()
    x = {0,0}
    while x[1] < 300 do
        x[1] = x[1] + 100
        box.session.push(x)
        fiber.sleep(100)
    end
    return x
end
box.space.tester:on_replace(
    on_replace_callback
)
```

2. 2 terminal:
```python
res = conn_pool.insert(
    'tester',
    (1, 'Mike'),
    mode=tarantool.Mode.PREFER_RO,
    on_push=callback,
    on_push_ctx=callback_res,
)
```

3. 1 terminal

Change a schema.

4. Wait ~100 seconds

Reply:
Actually, what behavior did we expect in case of receiving a SchemaReloadException?

Reply:
Sorry, but I couldn't reproduce the bug when changing the schema on the server in the process of receiving push:

I was making a delay in the client callback `callback(data, on_push_ctx=[])` with `time.sleep()`, and at that moment I was changing the scheme on the server with `box.space.tester:format()`, but there were no errors during push receiving

Reply:
To be honest, I also can't reproduce this one. I think we may just omit the testing for now and implement some adequate processing just in case it actually may happen.

After giving it some thought, I decided that Oleg's advice "If SchemaReloadException was thrown in the middle, just let it through" sounds like a good one. So I think we should use something like this
```python
response = None
while True:
    try:
        self._socket.sendall(bytes(request))
        response = request.response_class(self, self._read_response())
        break
    except SchemaReloadException as e:
        self.update_schema(e.schema_version)
        continue

while response._code == IPROTO_CHUNK:
    if on_push is not None: 
        if callable(on_push):
            on_push(response._data, on_push_ctx)
        else:
            raise TypeError('The on_push callback must be callable')
    response = request.response_class(self, self._read_response())

return response
```

and it should be alright if one manages to really get schema reload in the middle of the process.

Reply:
Thanks! Done